File: sdk/core/ios/Sources/Adapter/AdapterContract.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#endif

// MARK: - Data Contracts

/// Publisher-supplied vendor credentials injected at runtime. Keep on-device.
/// Publisher-supplied vendor credentials injected at runtime. Keep on-device and never log.
public struct AdapterCredentials: Codable, Equatable, Sendable {
    public let key: String
    public let secret: String?
    public let appId: String?
    public let accountIds: [String: String]?

    /// Returns a redacted/hashed view for telemetry/debugging (secrets omitted).
    public func redacted() -> [String: String] {
        var result: [String: String] = [:]
        result["key_hash"] = Self.hashValue(key)
        if let appId { result["app_id_hash"] = Self.hashValue(appId) }
        if let accounts = accountIds, !accounts.isEmpty {
            result["account_ids"] = accounts.keys.sorted().joined(separator: ",")
        }
        return result
    }

    private static func hashValue(_ input: String) -> String {
        // lightweight SHA256 hex for non-cryptographic labeling; secrets never emitted raw
        if let data = input.data(using: .utf8) {
            #if canImport(CryptoKit)
            return SHA256.hash(data: data).compactMap { String(format: "%02x", $0) }.joined()
            #else
            return String(input.hashValue)
            #endif
        }
        return ""
    }
}

/// Optional knobs the controller can hand to adapters per placement.
public struct AdapterOptions: Codable, Equatable, Sendable {
    public let startMuted: Bool?
    public let testMode: Bool?
    public let bidFloorMicros: Int64?
}

public enum PartnerRegion: String, Codable, Sendable {
    case us, eu, apac, cn, global
}

/// Normalized privacy metadata forwarded to vendors.
public struct ConsentState: Codable, Equatable, Sendable {
    public let gdprApplies: Bool?
    public let iabTCFv2: String?
    public let iabUSPrivacy: String?
    public let coppa: Bool
    public let attStatus: ATTStatus
    public let limitAdTracking: Bool
    public let privacySandboxOptIn: Bool?
    public let advertisingId: String?
    public let appSetId: String?
}

public enum ATTStatus: String, Codable, Sendable {
    case authorized, denied, notDetermined, restricted
}

/// Canonical init payload mirrored across Android, iOS, and Unity.
public struct PartnerAdapterConfig: Codable, Equatable, Sendable {
    public let partner: String
    public let credentials: AdapterCredentials
    public let placements: [String: String]
    public let privacy: ConsentState
    public let region: PartnerRegion?
    public let options: AdapterOptions?
}

public struct DeviceMeta: Codable, Equatable, Sendable {
    public let os: String
    public let osVersion: String
    public let model: String
}

public struct UserMeta: Codable, Equatable, Sendable {
    public let ageRestricted: Bool
    public let consent: ConsentState
    public let advertisingId: String?
    public let appSetId: String?
}

public struct NetworkMeta: Codable, Equatable, Sendable {
    public let ipPrefixed: String
    public let uaNormalized: String
    public let connType: ConnectionType
}

public enum ConnectionType: String, Codable, Sendable {
    case wifi, cell, other
}

public struct ContextMeta: Codable, Equatable, Sendable {
    public let orientation: Orientation
    public let sessionDepth: Int
}

public enum Orientation: String, Codable, Sendable {
    case portrait, landscape
}

public struct AuctionMeta: Codable, Equatable, Sendable {
    public let floorsMicros: Int64?
    public let sChain: String?
    public let sellersJsonOk: Bool?
}

/// Per-request context (device, user, auction) supplied to adapters.
public struct RequestMeta: Codable, Equatable, Sendable {
    public let requestId: String
    public let device: DeviceMeta
    public let user: UserMeta
    public let net: NetworkMeta
    public let context: ContextMeta
    public let auction: AuctionMeta
}

/// Token referencing a single-use creative returned by an adapter.
public struct AdHandle: Codable, Equatable, Sendable {
    public let id: String
    public let adType: AdType
    public let partnerPlacementId: String?
    public let createdAtMs: Int64
    
    public init(id: String, adType: AdType, partnerPlacementId: String?, createdAtMs: Int64 = Clock.shared.nowMillis()) {
        self.id = id
        self.adType = adType
        self.partnerPlacementId = partnerPlacementId
        self.createdAtMs = createdAtMs
    }
}

/// Successful load response; ttlMs governs cache freshness.
public struct LoadResult: Codable, Sendable {
    public let handle: AdHandle
    public let ttlMs: Int
    public let priceMicros: Int64?
    public let currency: String?
    public let partnerMeta: [String: AnyCodable]
}

/// Outcome of adapter initialization.
public struct InitResult: Codable, Sendable {
    public let success: Bool
    public let error: AdapterError?
    public let partnerMeta: [String: AnyCodable]
}

public struct PaidEvent: Codable, Equatable, Sendable {
    public let valueMicros: Int64
    public let currency: String
    public let precision: PaidEventPrecision
    public let partner: String
    public let partnerUnitId: String?
    public let lineItemId: String?
    public let creativeId: String?
}

public enum PaidEventPrecision: String, Codable, Sendable { case publisher, estimated }

// MARK: - Error Taxonomy

public enum AdapterErrorCode: String, Codable, Sendable {
    case noFill
    case timeout
    case networkError
    case status4xx
    case status5xx
    case belowFloor
    case error
    case circuitOpen
    case config
    case noAdReady
}

/// Normalized failure mapped from vendor-specific errors.
public struct AdapterError: Error, Codable, Equatable, Sendable {
    public let code: AdapterErrorCode
    public let detail: String
    public let vendorCode: String?
    public let recoverable: Bool

    /// Normalizes vendor reasons into canonical strings for parity with Android.
    public func normalizedReason() -> String {
        switch code {
        case .noFill: return "no_fill"
        case .timeout: return "timeout"
        case .networkError: return "network_error"
        case .status4xx: return "status_4xx"
        case .status5xx: return "status_5xx"
        case .belowFloor: return "below_floor"
        case .circuitOpen: return "circuit_open"
        case .config: return "config"
        case .noAdReady: return "no_ad_ready"
        case .error: return "error"
        }
    }
}

// MARK: - Callbacks

public enum CloseReason: String, Codable, Sendable { case completed, skipped, dismissed }

public protocol ShowCallbacks: AnyObject {
    func onImpression(meta: [String: Any]?)
    func onPaidEvent(_ event: PaidEvent)
    func onClick(meta: [String: Any]?)
    func onClosed(reason: CloseReason)
    func onError(_ error: AdapterError)
}

public protocol RewardedCallbacks: ShowCallbacks {
    func onRewardVerified(rewardType: String, rewardAmount: Double)
}

public protocol BannerCallbacks: ShowCallbacks {
    func onViewAttached()
    func onViewDetached(reason: CloseReason)
}

public struct AdSize: Codable, Equatable, Sendable {
    public let width: Int
    public let height: Int
    public static let banner320x50 = AdSize(width: 320, height: 50)
}

// MARK: - Core Adapter Protocol

/// Shared adapter surface enforced across all ApexMediation platforms.
public protocol AdNetworkAdapterV2 {
    func initAdapter(config: PartnerAdapterConfig, timeoutMs: Int) -> InitResult
    func loadInterstitial(placementId: String, meta: RequestMeta, timeoutMs: Int) -> LoadResult
    func showInterstitial(handle: AdHandle, viewController: AnyObject, callbacks: ShowCallbacks)
    func loadRewarded(placementId: String, meta: RequestMeta, timeoutMs: Int) -> LoadResult
    func showRewarded(handle: AdHandle, viewController: AnyObject, callbacks: RewardedCallbacks)
    func loadBanner(placementId: String, size: AdSize, meta: RequestMeta, timeoutMs: Int) -> LoadResult
    func attachBanner(handle: AdHandle, hostView: AnyObject, callbacks: BannerCallbacks)
    func isAdReady(handle: AdHandle) -> Bool
    func expiresAt(handle: AdHandle) -> Int64
    func invalidate(handle: AdHandle)
}

File: sdk/core/android/src/main/kotlin/contract/AdapterContract.kt
package com.rivalapexmediation.sdk.contract

import com.rivalapexmediation.sdk.models.AdType
import com.rivalapexmediation.sdk.util.ClockProvider
import java.security.MessageDigest

/**
 * Canonical adapter contract derived from backend Adapters.md spec.
 * These data classes and interfaces are platform-agnostic snapshots that
 * mediate between host SDK controllers and network-specific adapters.
 */

// ----- Data Contracts -----

/**
 * Publisher-supplied vendor credentials injected at runtime.
 * All fields are BYO secrets and must stay on-device.
 */
data class AdapterCredentials(
    val key: String,
    val secret: String? = null,
    val appId: String? = null,
    val accountIds: Map<String, String>? = null
) {
    // Redact secrets by default if ever stringified or logged.
    override fun toString(): String = buildString {
        append("AdapterCredentials(key=")
        append(hash(key))
        append(", secret=")
        append(secret?.let { "***" } ?: "null")
        append(", appId=")
        append(appId?.let { hash(it) } ?: "null")
        append(", accountIds=")
        append(accountIds?.mapValues { (_, v) -> hash(v) })
        append(")")
    }

    private fun hash(value: String): String = try {
        val digest = MessageDigest.getInstance("SHA-256").digest(value.toByteArray())
        digest.take(6).joinToString("") { byte -> "%02x".format(byte) }
    } catch (_: Exception) {
        "hash_err"
    }
}

/** Adapter tuning knobs supplied by the controller per placement. */
data class AdapterOptions(
    val startMuted: Boolean? = null,
    val testMode: Boolean? = null,
    val bidFloorMicros: Long? = null
)

enum class PartnerRegion { US, EU, APAC, CN, GLOBAL }

/** Normalized privacy metadata forwarded to every adapter request. */
data class ConsentState(
    val gdprApplies: Boolean? = null,
    val iabTcfV2: String? = null,
    val iabUsPrivacy: String? = null,
    val coppa: Boolean = false,
    val attStatus: AttStatus = AttStatus.NOT_DETERMINED,
    val limitAdTracking: Boolean = false,
    val privacySandboxOptIn: Boolean? = null,
    val advertisingId: String? = null,
    val appSetId: String? = null
)

enum class AttStatus { AUTHORIZED, DENIED, NOT_DETERMINED, RESTRICTED }

/** Canonical init payload passed to each adapter implementation. */
data class AdapterConfig(
    val partner: String,
    val credentials: AdapterCredentials,
    val placements: Map<String, String>,
    val privacy: ConsentState,
    val region: PartnerRegion? = null,
    val options: AdapterOptions? = null
)

data class DeviceMeta(
    val os: String,
    val osVersion: String,
    val model: String
)

data class UserMeta(
    val ageRestricted: Boolean,
    val consent: ConsentState,
    val advertisingId: String? = null,
    val appSetId: String? = null
)

data class NetworkMeta(
    val ipPrefixed: String,
    val uaNormalized: String,
    val connType: ConnectionType
)

enum class ConnectionType { WIFI, CELL, OTHER }

data class ContextMeta(
    val orientation: Orientation,
    val sessionDepth: Int
)

enum class Orientation { PORTRAIT, LANDSCAPE }

data class AuctionMeta(
    val floorsMicros: Long? = null,
    val sChain: String? = null,
    val sellersJsonOk: Boolean? = null
)

/** Per-request context (device, user, auction) shared with adapters. */
data class RequestMeta(
    val requestId: String,
    val device: DeviceMeta,
    val user: UserMeta,
    val net: NetworkMeta,
    val context: ContextMeta,
    val auction: AuctionMeta
)

/** Successful load response returned by an adapter. */
data class LoadResult(
    val handle: AdHandle,
    val ttlMs: Int,
    val priceMicros: Long? = null,
    val currency: String? = null,
    val partnerMeta: Map<String, Any?> = emptyMap()
)

data class InitResult(
    val success: Boolean,
    val error: AdapterError? = null,
    val partnerMeta: Map<String, Any?> = emptyMap()
)

/** Token representing a single-render creative. Single-use by design. */
data class AdHandle(
    val id: String,
    val adType: AdType,
    val partnerPlacementId: String? = null,
    val createdAtMs: Long = ClockProvider.clock.monotonicNow()
)

data class PaidEvent(
    val valueMicros: Long,
    val currency: String,
    val precision: PaidEventPrecision,
    val partner: String,
    val partnerUnitId: String? = null,
    val lineItemId: String? = null,
    val creativeId: String? = null
)

enum class PaidEventPrecision { PUBLISHER, ESTIMATED }

// ----- Error taxonomy -----

enum class ErrorCode {
    NO_FILL,
    TIMEOUT,
    NETWORK_ERROR,
    STATUS_4XX,
    STATUS_5XX,
    BELOW_FLOOR,
    ERROR,
    CIRCUIT_OPEN,
    CONFIG,
    NO_AD_READY
}

/** Normalized error returned by adapters; maps vendor failures to the taxonomy. */
sealed class AdapterError(open val code: ErrorCode, open val detail: String, open val vendorCode: String? = null) : Exception("[$code] $detail") {
    data class Fatal(override val code: ErrorCode, override val detail: String, override val vendorCode: String? = null): AdapterError(code, detail, vendorCode)
    data class Recoverable(override val code: ErrorCode, override val detail: String, override val vendorCode: String? = null): AdapterError(code, detail, vendorCode)

    /**
     * Normalized reason string for telemetry/debug surfaces.
     * This keeps Console debugger and adapters aligned on a shared taxonomy.
     */
    fun normalizedReason(): String = when (code) {
        ErrorCode.NO_FILL -> "no_fill"
        ErrorCode.TIMEOUT -> "timeout"
        ErrorCode.NETWORK_ERROR -> "network_error"
        ErrorCode.STATUS_4XX -> "status_4xx"
        ErrorCode.STATUS_5XX -> "status_5xx"
        ErrorCode.BELOW_FLOOR -> "below_floor"
        ErrorCode.CIRCUIT_OPEN -> "circuit_open"
        ErrorCode.CONFIG -> "config_error"
        ErrorCode.NO_AD_READY -> "no_ad_ready"
        ErrorCode.ERROR -> "error"
    }
}

// ----- Callback contracts -----

interface ShowCallbacks {
    fun onImpression(meta: Map<String, Any?> = emptyMap())
    fun onPaidEvent(event: PaidEvent)
    fun onClick(meta: Map<String, Any?> = emptyMap())
    fun onClosed(reason: CloseReason)
    fun onError(error: AdapterError)
}

interface RewardedCallbacks : ShowCallbacks {
    fun onRewardVerified(rewardType: String, rewardAmount: Double)
}

interface BannerCallbacks : ShowCallbacks {
    fun onViewAttached()
    fun onViewDetached(reason: CloseReason)
}

enum class CloseReason { COMPLETED, SKIPPED, DISMISSED }

// ----- Core adapter interface (v2) -----

/**
 * Shared adapter surface enforced on Android/iOS/Unity.
 * Implementations must obey threading, timeout, and single-use handle rules.
 */
interface AdNetworkAdapterV2 {
    fun init(config: AdapterConfig, timeoutMs: Int): InitResult
    fun loadInterstitial(placementId: String, meta: RequestMeta, timeoutMs: Int): LoadResult
    fun showInterstitial(handle: AdHandle, viewContext: Any, callbacks: ShowCallbacks)
    fun loadRewarded(placementId: String, meta: RequestMeta, timeoutMs: Int): LoadResult
    fun showRewarded(handle: AdHandle, viewContext: Any, callbacks: RewardedCallbacks)
    fun loadBanner(placementId: String, size: AdSize, meta: RequestMeta, timeoutMs: Int): LoadResult
    fun attachBanner(handle: AdHandle, bannerHost: Any, callbacks: BannerCallbacks)
    fun isAdReady(handle: AdHandle): Boolean
    fun expiresAt(handle: AdHandle): Long
    fun invalidate(handle: AdHandle)
}

data class AdSize(val width: Int, val height: Int)